### PR TITLE
Bug fix for wrong sign of curl in maxwell.h

### PR DESCRIPTION
**Pull request by orlando-tian**

### CODE REVIEW

**Comment by bangerth** [comment]
Fixes #2525.

**Comment by bangerth** [comment]
@guidokanschat -- can you take a look at this?

**Comment by kronbichler** [comment]
This looks good. Is there any test that could help us assert that we do things right now (and that the stay right)?

**Comment by orlando-tian** [comment]
@kronbichler Because I'm a new user of Github and dealii, would you please tell me where I can find the coding convention required by dealii project for writing a test? Thank you!

**Comment by kronbichler** [comment]
For general info about tests, please see here:
https://dealii.org/developer/developers/testsuite.html

For the more specific task of setting up a test here, I would propose you set up a simple case on a mesh with just two elements (to also get the face term) and output the matrix that gets assembled into `deallog` rather than `cout`. The output file that comes out of this process would then be stored as test_name.output and added to `tests/integrators`. You can have a look at the various files in `tests/integrators` for examples of similar tests (but using a different integrator).

Thanks a lot for taking the time for writing a test!

**Comment by agrayver** [comment]
Although I haven't used this functionality myself, this does look like an issue to me with this patch fixing it correctly. 

@orlando-tian were you able to check with some analytical function that the sign is correct now?

**Comment by orlando-tian** [comment]
@agrayver Thanks for your suggestion! I'll first get familiar with how to generate a test code, then write a test for it.

**Comment by orlando-tian** [comment]
@bangerth @kronbichler @agrayver I've written a test code for curl integrator in maxwell.h, which is compared with analytical solution. Would you please help review it and provide your suggestions and comments? Thank you!

File: tests/integrators/maxwell_curl.mac [hunk 1]
```diff
@@ -0,0 +1,190 @@
+/*
+  Copyright (C) 2012 - 2015 by the deal.II authors
```
Comment (kronbichler) [inline tests/integrators/maxwell_curl.mac]: @orlando-tian Did you add this file by mistake? I cannot make sense of it in the current deal.II testing framework (even though it looks interesting)...

File: tests/integrators/maxwell_curl.cc [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+template <int dim>
+void make_grid(Triangulation<dim> &tr)
+{
+  GridGenerator::subdivided_hyper_cube(tr, 1, -1., 1.);
+
+  std::ofstream out ("grid.eps");
+  GridOut grid_out;
+  grid_out.write_eps (tr, out);
+  std::cout << "Grid written to grid.eps" << std::endl;
```
Comment (kronbichler) [inline tests/integrators/maxwell_curl.cc]: Could you please remove the grid output? We don't need that in the test suite.

File: tests/integrators/maxwell_curl.cc [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  const unsigned int dofs_per_cell = fe.dofs_per_cell;
+
+  FullMatrix<double> curl_curl_check(dofs_per_cell, dofs_per_cell);
+  FullMatrix<double> curl_check(dofs_per_cell, dofs_per_cell);
+  FullMatrix<double> nitsche_curl_check(dofs_per_cell, dofs_per_cell);
+
+  curl_curl_check = 0;
+  curl_check = 0;
+  nitsche_curl_check = 0;
+
```
Comment (kronbichler) [inline tests/integrators/maxwell_curl.cc]: The automatic indentation check complains that you have a non-empty line 80. Could you please fix that? (I think that is the only issue. In the future, you can avoid problems by running `astyle`, version 2.04, on the code you want to commit, see here https://dealii.org/developer/doxygen/deal.II/CodingConventions.html - the simple way is to run the script `contrib/utilities/indent` that runs astyle on all files.)

**Comment by davydden** [comment]
there is a minor indentation issue. You can install Astyle 2.04 and run `./contrib/utilities/indent` or check the `Travis CI` output and fix it by hand.

**Comment by kronbichler** [comment]
@orlando-tian Could you please add a short remark about the bug fix in the `doc/news/changes.h` file about your fix in the section `Specific improvements`? Thanks a lot for the test.

**Comment by agrayver** [comment]
@orlando-tian indeed looks good to me (subject to provided comments), just a little bit surprised you're using this functionality with FE_Q element (given that it's not curl-conforming), but it's unrelated ;)

**Comment by bangerth** [comment]
Oh, `FE_Q` is H_curl conforming alright -- if it's H^1 conforming, it must also be H_curl conforming. It just isn't dense in H_curl ;-)

File: tests/integrators/maxwell_curl.mac (hunk 1 above)
Comment (orlando-tian) [inline tests/integrators/maxwell_curl.mac]: @kronbichler OK, I've removed this file.

File: tests/integrators/maxwell_curl.cc (hunk 3 above)
Comment (orlando-tian) [inline tests/integrators/maxwell_curl.cc]: @kronbichler It's done.

**Comment by orlando-tian** [comment]
@bangerth Thanks for your explanation!

**Comment by kronbichler** [comment]
Now we are almost there. :+1:

1. Could you please rebase your branch to master - there is a conflict in `doc/news/changes.h` that needs to be resolved manually because another patch also added a line there.

2. Could you please squash your commits into one single commit?

Thanks!

**Comment by orlando-tian** [comment]
@kronbichler Thanks for your guidance! Because I just switched from CVS to git and I'm still not so familiar with git operations, would you please tell me how should I perform rebase and squash?

**Comment by kronbichler** [comment]
For the rebase, go to your local copy, switch to the master branch by `git checkout master`, do `git pull` to get the latest upgrades. Then switch back to your branch, I assume it is called `git checkout orlando-tian-contrib`. Then you call `git rebase master`. This should do the rebase and give you a merge conflict in `doc/news/changes.h`. Then, modify that file by hand by removing the merge markers and make sure your message as well as the other one in the conflict line agrees. Then, on the command line, you do `git add doc/news/changes.h` and then `git rebase --continue`.

As for the squashing: To merge all commits into one, first run `git rebase -i HEAD~5` (assuming that you have 5 commits). Then, you will get an editor that lists all your changes. In the window you should find five lines with 'pick' in the first column each. You simply change that word to squash on all lines but the first, which says that the later commits should be merged with the first one. You will then get a chance to write a new commit message that summarizes all five previous one, and then you're done.

Finally, do `git push [path to your online branch] --force`, the important difference to what you did before being the `force` keyword that allows to remove older commits.

**Comment by orlando-tian** [comment]
@kronbichler Thanks for your help! I've finished the rebase and squash according to your guidance and I hope my operation is correct.

**Comment by kronbichler** [comment]
Great!

**Comment by bangerth** [comment]
@orlando-tian -- thanks from me as well. Learning how a particular project wants their patches, and learning how to work with git and github is always a bit of an orduous process, and I'm glad you stuck with @kronbichler 's requests. It will get easier with your next patch requests, since you will already be able to anticipate what people may ask you for!

Thanks again!

**Comment by orlando-tian** [comment]
@bangerth You're welcome! I really learned a lot through this process. Thank you all! I hope to contribute more to deal.II in the future.

**Comment by bangerth** [comment]
@orlando-tian -- we look forward to more contributions! :-)

**Pull request merged**